### DISCUSSION TITLE
How to set targets to create ggplot graphs for different groups?

### DISCUSSION THREAD

**Discussion by leonardochalhoub** [ropensci/targets#609] [Help] [answered]
I got stuck on the list( ) part, couldnt generate it, although it seems easy. Could somebody help me please?

**Comment by wlandau** [comment, marked as answer]
Would you please elaborate on what you mean and provide some simple example code? I do not understand what you are trying to do or what a favorable outcome would look like.

**Reply by leonardochalhoub** [reply]
hello, thanks for answering and have a great sunday.
below is the code I'm trying to write - but I cant get it to work generating plots.

It works step by step perfectly until last  tar_target( name = graphs*

in your example that I watched, you make linear regressions at this point.
in that step, I actually want to generate different graphs using ggplot, generatin a file with different names for each group / Branch.

-
quick alignment: nothing special below. Using geobr package + ggplot2, I want to plot maps using database x, a regular database where variable X is a percent by region. I want to plot "variable X" by a group called Channel (I branch over it).

library(geobr)
library(targets)
library(tarchetypes)
library(ggplot2)
library(dplyr)

tar_option_set(packages = c("tidyverse", "broom", "ggplot2"))

states <- read_state(year=2019)

x <- readr::read_rds("~/database.rds") %>% 
  left_join(.,states, by= c("UF"="abbrev_state"))

list(
  tar_target(
    name = DATA,
    command = x
  ),
  tar_group_by(
    name = grouped_Channel,
    command = DATA,
    Channel
  ),
  tar_target(
    name = grouped_branches,
    command = grouped_Channel,
    pattern = map(grouped_Channel)
  ),
  tar_target(
    name = graphs,
    command = tibble(
      tar_group = grouped_branches$Channel %>% unique(),
      mode = list(
          tar_target(
            grouped_branches, {
              gg <-  ggplot() +
                geom_sf(data=grouped_branches, aes(fill= `variable X`))+
                viridis::scale_fill_viridis(name = "", discrete = F, 
                                            labels=scales::label_percent(
                                              accuracy = 1,
                                              scale = 100,
                                              prefix = "",
                                              suffix = "%",
                                              big.mark = " ",
                                              decimal.mark = ".",
                                              trim = TRUE
                                              ),
                                            limits=c(min(grouped_branches$`variable X`),max(grouped_branches$`variable X`)),
                                            option = "E",
                                            direction = -1)
              ggsave(gg, "~/grouped_branches{Channel}.png")
            
            },
            format = "file"
          )
        ),
      
      pattern = map(grouped_branches)
       
        )
      )
    
    )

**Reply by wlandau** [reply]
Looks like you have a target that calls another target: `tar_target(graphs, tibble(..., tar_target(grouped_branches, ...)))`. I think that's the issue. 

In the future, please post simpler code and format it so other people can read it more easily. The `styler` package and https://style.tidyverse.org/ can help. Also, please post runnable code. I cannot run what you posted because it depends on a `~/database.rds` file that I do not have. I suggest you have a look at https://reprex.tidyverse.org/articles/reprex-dos-and-donts.html.

**Reply by wlandau** [reply]
Here is a simpler `_targets.R` file that does work and other people can run. Might help you.

```r
# _targets.R file
library(targets)
library(tarchetypes)

tar_option_set(packages = "tidyverse")

list(
  tar_group_by(
    name = data,
    command = mtcars,
    cyl
  ),
  tar_target(
    name = graphs,
    command = {
      gg <- ggplot(data) + geom_point(aes(x = wt, y = mpg))
      ggsave(sprintf("plot%s.png", data$cyl[1]), gg)
    },
    pattern = map(data)
  )
)
```

**Reply by leonardochalhoub** [reply]
Thank you, it is working perfectly. I will learn to use reprex, very nice indeed.
Let me please ask you for a last help

Let's say it is working, with cyl (from the previous example) having 3 types, so 3 graphics generated. If I wanted to plot with another grouping variable, like cyl AND another (with, let's say, 2 types). I wish to plot all possible graphs (6). 
I did some tests but I couldnt perform it -> writing "cyl, new variable", it ran, but I couldn't find the correct way of labeling the files on ggsave line. Thank you in advance.

Targets is an awesome package, thank you for this great work.

**Reply by wlandau** [reply]
`tar_group_by()` can group by more than one column, so you can write something like `tar_group_by(data, mtcars, cyl, drat)`. To create the correct path for `ggsave()`, you can use `sprintf()` or `paste()` or `paste0()` to construct the file name, e.g. `sprintf("plot-%s-%s.png", data$cyl[1], data$drat[1])`.
